### PR TITLE
Update AzVolumeAttachment controller replica management logic

### LINKED ISSUE #21: Potential Replica Management Inconsistency Issue
**Is your feature request related to a problem?/Why is this needed**
With the current AzVolumeAttachment controller design, there is a potential risk for inconsistency that arises from using cached reader. Consider the following scenario:

1. Assume we have AzVolumeAttachment A and B for volume C.
2. A and B both gets deleted.
3. A's reconciliation worker thread acquires the lock first and starts replica management operation for volume C. B's worker threads wait for the lock to be released.
4. A's worker thread sees that there are are 2 less AzVolumeAttachments for volume C than desired.
5. A's worker thread creates 2 replica attachments for volume C, and releases the lock.
6. B's worker thread acquires the lock and begins replica management for volume C as well.
7. Cached reader hasn't synced yet and B's worker thread gets stale cache where it doesn't yet see the creation of 2 replicas from previous replica management.
8. B's worker thread creates 2 more replica attachment for volume C.
9. 4 replicas were created instead of 2.

Currently, there is a separate sync operation running on a separate thread that is triggered every 120 seconds to correct these inconsistencies but this is not ideal.

**Describe the solution you'd like in detail**

Currently, `ManageReplicas` function will create or delete `n` replicas based on diff between the `maxMountReplicaCount + 1` and `currentAttachmentCount`.

And `ManageReplica` is called in two cases:
1. after attaching a volume to a node (if it is a primary attachment)
2. after deleting a AzVolumeAttachment object

The discussed inconsistency can be overcome, if `ManageReplicas` is triggered upon every creation (currently triggered only when it is a creation of primary attachment) and deletion event of AzVolumeAttachment object (this will allow the number of replicas to eventually converge) and was capped to create or delete at most 1 replica at a time (this will reduce the bandwidth of fluctuations until conversion), but this could be a bit heavier than the current design as it would inherently require more `ManageReplica` calls and loops (one loop to create n replicas vs. n loops to create n replicas)

It would look something like this:
```
        var azVolumeAttachment v1alpha1.AzVolumeAttachment
	err := r.client.Get(ctx, request.NamespacedName, &azVolumeAttachment)
	// this is a deletion event
	if errors.IsNotFound(err) {
                r.ManageReplica()
		return reconcile.Result{}, nil
	}
...
	// this is a creation event
	if azVolumeAttachment.Status == nil {
               r.TriggerAttach()
               r.ManageReplica()
               return reconcile.Result{}, nil
       }
```

Creation scenario:
```
maxMountReplicaCount = 2
```
1. `ControllerPublishVolume` creates primary AzVolumeAttachment A for volume C.
3. `A.Status == nil`, so controller attaches A and calls `ManageReplica`
3. `currentAttachmentCount = 1` and `maxMountReplicaCount + 1 = 3`
4.  `currentAttachmentCount < maxMountReplicaCount + 1`, so create replica B
4. `B.Status == nil`, so controller attaches B and calls `ManageReplica`
5. `currentAttachmentCount = 2` and `maxMountReplicaCount + 1 = 3`
6. `currentAttachmentCount < maxMountReplicaCount + 1`, so create replica C
7. `C.Status == nil`, so controller attaches C and calls `ManageReplica`
8. `currentAttachmentCount = 3` and `maxMountReplicaCount + 1 = 3`
9. `currentAttachmentCount == maxMountReplicaCount + 1`, so stop

Inconsistencies can happen but will converge as below:

Deletion scenario:
1. AzVolumeAttachment A and B deleted.
2. `A` is not found, so call `ManageReplica`
3. `currentAttachmentCount = 1` and `maxMountReplicaCount + 1 = 3`
4. `currentAttachmentCount < maxMountReplicaCount + 1`, so create replica D
5. `B` is not found, so call `ManageReplica`
6. Cached reader may not know about `D`'s creation so can think that `currentAttachmentCount = 1` and `maxMountReplicaCount + 1 = 3`
7. `currentAttachmentCount < maxMountReplicaCount + 1`, so create replica E
8. `D.Status == nil`, so controller attaches D and calls `ManageReplica`
9. Cached Reader might not know about `E`'s creation, so can think that `currentAttachmentCount = 2` and `maxMountReplicaCount + 1 = 3`
9. `currentAttachmentCount < maxMountReplicaCount + 1`, so create replica F
10.  `E.Status == nil`, so controller attaches E and calls `ManageReplica`
11. Cached Reader definitely knows about both `D` and `E`'s creation and may not know about `F`'s creation, so can think that `currentAttachmentCount = 3` and `maxMountReplicaCount + 1 = 3` 
12. `currentAttachmentCount == maxMountReplicaCount + 1`, so stop
13. `F.Status == nil`, so controller attaches F and calls `ManageReplica`
14. Cached Reader definitely knows about `D`, `E`, and `F`'s creation, so `currentAttachmentCount = 4` and `maxMountReplicaCount + 1 = 3`
15. `currentAttachmentCount > maxMountReplicaCount + 1`, so delete a replica
16. Deletion of a replica triggers another `ManageReplicas`, but it knows that  `currentAttachmentCount == maxMountReplicaCount + 1`, so will stop.

Note that the ordering of who acquires lock first can change but the main point is that whoever acquires the lock last will be guaranteed to know about the creations and deletions of `AzVolumeAttachment` objects that triggered the calls that came before it. That said, inconsistencies can still happen but it will eventually converge.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by sunpa93**
**What type of PR is this?**

/kind design

**What this PR does / why we need it**:

updates replica management logic for AzVolumeAttachment controller.
adds convergence test to controller test suite

**Which issue(s) this PR fixes**:

Fixes #21 

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

this PR requires further investigation on whether this logic update is going to be beneficial performance-wise.

**Release note**:
```
none
```

### CODE REVIEW

File: test/e2e/controller_test.go [hunk 1]
```diff
@@ -214,6 +214,46 @@ var _ = ginkgo.Describe("Controller", func() {
...
+				}
+			}
+			time.Sleep(time.Duration(30) * time.Second)
+
+			// check if the number of replicas have properly converged
+			err = testAzAtt.WaitForReplicas(replicaCount, time.Duration(5)*time.Minute)
+			framework.ExpectNoError(err)
+
+			// clean-up
+			testAzAtt.Cleanup()
```
Comment (landreasyan) [inline test/e2e/controller_test.go:254]: I suggest using `defer` to call cleanup right after object is created. That ensures that we don't forget to call cleanup at the end of the function.

Comment (sunpa93) [reply test/e2e/controller_test.go:254]: sounds good :)

File: pkg/controller/azvolumeattachment.go [hunk 2]
```diff
@@ -684,17 +678,19 @@ func NewAzVolumeAttachmentController(ctx context.Context, mgr manager.Manager, a
...
+	err = reconciler.SyncAll(ctx)
+	if err != nil {
+		klog.Warningf("failed to complete initial AzVolumeAttachment sync: %v", err)
+	}
 
 	klog.V(2).Info("AzVolumeAttachment Controller successfully initialized.")
 	return nil
 }
+
+func min(a, b int) int {
```
Comment (landreasyan) [inline pkg/controller/azvolumeattachment.go:691]: Any reason we can't use Min in "math" package here?

Comment (sunpa93) [reply pkg/controller/azvolumeattachment.go:691]: `Min` in math package uses float64 and requires conversion so decided to define one for the sake of simplicity :)

File: pkg/controller/azvolumeattachment.go [hunk 3]
```diff
@@ -224,18 +222,17 @@ func (r *reconcileAzVolumeAttachment) SyncVolume(ctx context.Context, azVolume v
 
 	if desiredAttachmentCount > currentAttachmentCount {
 		klog.Infof("Create %d more replicas for volume (%s)", desiredAttachmentCount-currentAttachmentCount, azVolume.Spec.UnderlyingVolume)
-		if err = r.CreateReplicas(ctx, desiredAttachmentCount-currentAttachmentCount, azVolume.Spec.UnderlyingVolume, useCache); err != nil {
+		if err = r.CreateReplicas(ctx, min(defaultMaxReplicaUpdateCount, desiredAttachmentCount-currentAttachmentCount), azVolume.Spec.UnderlyingVolume, useCache); err != nil {
```
[the remark this answers is not in the document]
Comment (sunpa93) [reply pkg/controller/azvolumeattachment.go:225]: `defaultMaxReplicaUpdateCount` was created to define the maximum number of creation or deletion of AzVolumeAttachment objects in a single `ManageReplica` call.

When `defaultMaxReplicaUpdateCount == 1`, having `min(defaultMaxReplicaUpdateCount, ...)` would be trivial for the reason you mentioned, but if we decide to set `defaultMaxReplicaUpdateCount > 1`, let's say 2, we should not be creating or deleting 2 replicas at all instances. Instead, we should be creating or deleting only 1 replica if the delta between the desired and current is 1, and `min(defaultMaxReplicaUpdateCount, ...)` assures that.

[the remark this answers is not in the document]
Comment (edreed) [reply pkg/controller/azvolumeattachment.go:225]: Duh. I reversed `min` with `max`. :/ I'm not sure that we'll ever use a value other than 1, though. Let it stand, but add a comment to the constant describing its intended use.

[the remark this answers is not in the document]
Comment (sunpa93) [reply pkg/controller/azvolumeattachment.go:225]: Will do :)

**Review by edreed — APPROVED** [review]

**Pull request merged**